File: Beecrowd_Parte1/AmigosHabay2136.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 100
#define MAX_PARTICIPANTES 1000

// Função para comparar strings (para qsort)
int compare(const void *a, const void *b) {
    return strcmp(*(const char **)a, *(const char **)b);
}

int main() {
    char participantes[MAX_PARTICIPANTES][MAX_LENGTH];
    char participantes_no[MAX_PARTICIPANTES][MAX_LENGTH]; // Array para participantes que disseram "NO"
    char amigo[MAX_LENGTH];
    int index = 0;
    int index_no = 0; // Índice para o array de participantes que disseram "NO"
    int i;

    while (1) {
        char entrada[MAX_LENGTH];
        scanf("%s", entrada);

        if (strcmp(entrada, "FIM") == 0) {
            break;
        }

        char nome[MAX_LENGTH];
        char status[MAX_LENGTH];
        sscanf(entrada, "%s %s", nome, status);

        if (strcmp(status, "YES") == 0) {
            // Verifica se o nome já está na lista de participantes YES
            int encontrado = 0;
            for (i = 0; i < index; i++) {
                if (strcmp(participantes[i], nome) == 0) {
                    encontrado = 1;
                    break;
                }
            }
            if (!encontrado) {
                strcpy(participantes[index], nome);
                if (strlen(nome) > strlen(amigo)) {
                    strcpy(amigo, nome);
                }
                index++;
            }
        } else {
            // Verifica se o nome já está na lista de participantes NO
            int encontrado = 0;
            for (i = 0; i < index_no; i++) {
                if (strcmp(participantes_no[i], nome) == 0) {
                    encontrado = 1;
                    break;
                }
            }
            if (!encontrado) {
                strcpy(participantes_no[index_no], nome);
                index_no++;
            }
        }
    }

    // Ordena os participantes YES lexicograficamente
    qsort(participantes, index, sizeof(participantes[0]), compare);

    // Imprime os participantes YES
    printf("Participantes que disseram YES:\n");
    for (i = 0; i < index; i++) {
        printf("%s\n", participantes[i]);
    }

    // Imprime os participantes NO
    printf("\nParticipantes que disseram NO:\n");
    for (i = 0; i < index_no; i++) {
        printf("%s\n", participantes_no[i]);
    }

    // Imprime o amigo do Habay
    printf("\nAmigo do Habay:\n");
    printf("%s\n", amigo);

    return 0;
}

File: Beecrowd_Parte1/AmigosHabay2136.java
import java.util.HashSet;
import java.util.Set;
import java.util.Scanner;
import java.util.TreeSet;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        Set<String> yesSet = new TreeSet<>(); // Utiliza TreeSet para ordenar os nomes automaticamente
        Set<String> noSet = new TreeSet<>();  // Utiliza TreeSet para ordenar os nomes automaticamente
        String amigo = "";
        
        String[] entrada = sc.nextLine().split(" ");
        
        while (!entrada[0].equals("FIM")) {
            String nome = entrada[0];
            String status = entrada[1];
            
            if (status.equals("YES")) {
                yesSet.add(nome);
                if (nome.length() > amigo.length()) {
                    amigo = nome;
                }
            } else {
                noSet.add(nome);
            }
            
            entrada = sc.nextLine().split(" ");
        }
        
        for (String nome : yesSet) {
            System.out.println(nome);
        }
        
        for (String nome : noSet) {
            System.out.println(nome);
        }
        
        System.out.println("\nAmigo do Habay:");
        System.out.println(amigo);
        
        sc.close();
    }
}
